config/initializers/foreman.rb

```diff
@@ -32,17 +31,17 @@
   Foreman.settings.load_values
 end
 
-# load topbar
-Menu::Loader.load
-
-# clear our users topbar cache
-# The users table may not be exist during initial migration of the database
-TopbarSweeper.expire_cache_all_users if (User.table_exists? rescue false)
-
 Foreman::Plugin.initialize_default_registries
 Foreman::Plugin.medium_providers_registry.register MediumProviders::Default
 
 Rails.application.config.to_prepare do
+  # load topbar
+  Menu::Loader.load
```

Comment:
The menu disappears for me with this change.

Though it's wrong even conceptually I believe, `Menu::Manager` is in registries folder, which is autoload_once (never reloaded). This is probably wrong and we should reload menu on code reload, but we do not, so it would load the menu items multiple times.

So IMHO, this is a candidate for `before_initialize`, because it has already initialized the `Rails.autoloaders.once`, which holds the Menu constant. Check the [bootstrap.rb](https://github.com/rails/rails/blob/5a985451f938a35dfdde862615b3e6f20131ed16/railties/lib/rails/application/bootstrap.rb#L88)

Reply:
> The menu disappears for me with this change.

Really? Because it works for me fine (on vanilla foreman, no plugins)

Reply:
> Though it's wrong even conceptually I believe, Menu::Manager is in registries folder, which is autoload_once (never reloaded). This is probably wrong and we should reload menu on code reload, but we do not, so it would load the menu items multiple times.

Maybe we should move it to services and reload it on every change

Reply:
Yeah, when you add plugins into the mix it stop working, because the root menus got initialized from plugins too soon and Foreman won't reinitialize them properly.

> Maybe we should move it to services and reload it on every change

I believe we should, there is no real reason to initialize it soon, but that's a bit of a challange to do it correctly as there might be :dragon:s  :)
If you're up for implementing it I'm all for it, but putting it into `before_initialize` is acceptable workaround for me :)

Reply:
For now I put `Menu::Loader.load` to `before_initialize`, for follow up refactoring I created https://projects.theforeman.org/issues/33975 so if there will be time I'll look into it